README.md
# ndarray-heatmap

Renders ndarray into a canvas.

[![Travis build status](http://img.shields.io/travis/akngs/ndarray-heatmap.svg?style=flat)](https://travis-ci.org/akngs/ndarray-heatmap)
[![Code Climate](https://codeclimate.com/github/akngs/ndarray-heatmap/badges/gpa.svg)](https://codeclimate.com/github/akngs/ndarray-heatmap)
[![Test Coverage](https://codeclimate.com/github/akngs/ndarray-heatmap/badges/coverage.svg)](https://codeclimate.com/github/akngs/ndarray-heatmap)
[![Dependency Status](https://david-dm.org/akngs/ndarray-heatmap.svg)](https://david-dm.org/akngs/ndarray-heatmap)
[![devDependency Status](https://david-dm.org/akngs/ndarray-heatmap/dev-status.svg)](https://david-dm.org/akngs/ndarray-heatmap#info=devDependencies)

## Install

    npm install ndarray-heatmap

## Example

Plain old javascript:

    <script src="dist/ndarray-heatmap.min.js"></script>
    <script>
    window.addEventListener('load', function() {
        var shape = [100, 100];
        var data = new Array(shape[0]);
        for(var i = 0; i < shape[0]; ++i) {
            data[i] = new Array(shape[1]);
            for(var j = 0; j < shape[1]; ++j) {
                data[i][j] = j * shape[0] + i;
            }
        }

        var renderer = heatmap.heatmap()
            .data(data)
            .colorRange(['white', 'red']);
        var canvas = renderer();
        document.body.appendChild(canvas);
    </script>

ES6 with ndarray:

    import heatmap from 'ndarray-heatmap';
    import pack from 'ndarray-pack';

    var array = pack([
        [0, 0, 0, 0, 0, 0],
        [0, 1, 1, 1, 1, 0],
        [0, 1, 2, 2, 1, 0],
        [0, 1, 1, 1, 1, 0],
        [0, 0, 0, 0, 0, 0]
    ]);
    var renderer = heatmap()
        .data(array)
        .colorRange(['#000000', '#FFFFFF']);
    var canvas = renderer();

Or, you can provide existing canvas element:

    var canvas = document.createElement('canvas');
    renderer(canvas);

If you are not familiar with ``ndarray`` please read [scijs documentation](http://scijs.net/packages/).



CHANGELOG.md
### [0.0.1](https://github.com/akngs/ndarray-heatmap/releases/tag/v0.0.1)

- The first release


test/unit/ndarray-heatmap.js
import { heatmap } from '../../src/ndarray-heatmap';
import pack from 'ndarray-pack';
import unpack from 'ndarray-unpack';

describe('heatmap', () => {
  let hm;

  beforeEach(() => {
    hm = heatmap();
  });

  it('should accept plain JS array', () => {
    hm.data([[1, 2], [3, 4]]);
    expect(unpack(hm.data())).to.deep.equal([[1, 2], [3, 4]]);
  });

  it('should provide domain accessor', () => {
    // default value
    expect(hm.domain()).to.equal(null);

    // setter and getter
    expect(hm.domain([0, 10])).to.equal(hm);
    expect(hm.domain()).to.deep.equal([0, 10]);
  });

  it('should provide data accessor', () => {
    // default value
    expect(unpack(hm.data())).to.deep.equal([[0]]);

    // setter and getter
    expect(hm.data(pack([[1, 2], [3, 4]]))).to.equal(hm);
    expect(unpack(hm.data())).to.deep.equal([[1, 2], [3, 4]]);
  });

  it('should provide colorSteps accessor', () => {
    // default value
    expect(hm.colorSteps()).to.equal(256);

    // setter and getter
    expect(hm.colorSteps(10)).to.equal(hm);
    expect(hm.colorSteps()).to.equal(10);
  });

  it('should provide colorRange accessor', () => {
    // default value
    expect(hm.colorRange()).to.deep.equal(['#000000', '#FFFFFF']);

    // setter and getter
    expect(hm.colorRange(['#FF0000', '#00FFFF'])).to.equal(hm);
    expect(hm.colorRange()).to.deep.equal(['#FF0000', '#00FFFF']);
  });

  it('should reject non-2d arrays', () => {
    let flat = pack([1])
    let threed = pack([[[1]]])

    expect(() => hm.data(flat)).to.throw(/Invalid rank/);
    expect(() => hm.data(threed)).to.throw(/Invalid rank/);
  });

  if(typeof window !== 'undefined') {
    it('should use shape[0] as a height', () => {
      hm.data(pack([[1, 2, 3, 4], [5, 6, 7, 8], [9, 10, 11, 12]]));

      let canvas = hm();
      expect(canvas.width).to.equal(4);
      expect(canvas.height).to.equal(3);
    });
  }
});



src/ndarray-heatmap.js
import ndarray from 'ndarray';
import pack from 'ndarray-pack';
import cwise from 'cwise';
import { extent } from 'd3-array';
import { interpolateLab } from 'd3-interpolate';
import { rgb } from 'd3-color';

const renderToCanvas = cwise({
  args: [
    'index', 'array',
    'scalar', 'scalar', 'scalar', 'scalar', 'scalar'
  ],
  body: function(i, value, imgArray, colors, min, max, imgWidth) {
    let colorIndex = Math.round((value - min) / (max - min) * (colors.length - 1));
    let {r, g, b, a} = colors[colorIndex];
    let base = (i[0] * imgWidth + i[1]) * 4;
    imgArray[base] = r;
    imgArray[++base] = g;
    imgArray[++base] = b;
    imgArray[++base] = a || 255;
  }
});

function heatmap() {
  let data = ndarray(new Float64Array([0]), [1, 1]);
  let colorSteps = 256;
  let domain = null;
  let colorRange = ['#000000', '#FFFFFF'];

  function render(_) {
    let canvas = _ || document.createElement('canvas');
    canvas.width = data.shape[1];
    canvas.height = data.shape[0];

    let ctx = canvas.getContext('2d');
    let imgData = ctx.getImageData(0, 0, canvas.width, canvas.height);
    let imgArray = imgData.data;
    let [min, max] = domain || extent(data.data);
    let colorScale = interpolateLab(...colorRange);
    let colors = [];
    for(let i = 0; i < colorSteps; ++i) {
      colors.push(rgb(colorScale(i / (colorSteps - 1))));
    }
    renderToCanvas(data, imgArray, colors, min, max, canvas.width);
    ctx.putImageData(imgData, 0, 0);

    return canvas;
  }

  render.data = function(_) {
    if(!arguments.length) return data;

    // Convert plain JS array into ndarray
    _ = _.shape ? _ : pack(_);

    if(_.shape.length !== 2) throw new Error(`Invalid rank: ${_.shape.length}`);
    data = _;
    return render;
  }

  render.colorSteps = function(_) {
    return arguments.length ? (colorSteps = _, render) : colorSteps;
  }

  render.domain = function(_) {
    return arguments.length ? (domain = _, render) : domain;
  }

  render.colorRange = function(_) {
    return arguments.length ? (colorRange = _, render) : colorRange;
  }

  return render;
}

export {
  heatmap
};
